idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/frontend/api/fir/components/KtFirReferenceShortener.kt

```diff
@@ -295,60 +339,76 @@ private class ElementsToShortenCollector(
+                // the reference.
+                availableClassifier == null -> {
+                    // Caller indicates don't shorten if doing that needs importing more names. Hence, we just skip.
+                    if (option == ShortenOption.SHORTEN_IF_ALREADY_IMPORTED) continue
+                    return createElementToShorten(
+                        element,
+                        classId.asSingleFqName(),
+                        option == ShortenOption.SHORTEN_AND_IMPORT_ALL_IN_PARENT
+                    )
+                }
+                // The class with name `classId.shortClassName` happens to be the same class referenced by this qualified access.
+                availableClassifier.symbol == classId -> {
+                    // Respect caller's request to use star import, if it's not already star-imported.
+                    return when {
+                        availableClassifier.importKind == ImportKind.EXPLICIT && option == ShortenOption.SHORTEN_AND_IMPORT_ALL_IN_PARENT -> {
+                            createElementToShorten(element, classId.asSingleFqName(), true)
+                        }
+                        // Otherwise, just shorten it and don't alter import statements
+                        else -> createElementToShorten(element, null, false)
+                    }
+                }
+                // Allow using star import to overwrite members implicitly imported by default.
+                availableClassifier.importKind == ImportKind.DEFAULT_ON_DEMAND && option == ShortenOption.SHORTEN_AND_IMPORT_ALL_IN_PARENT -> {
+                    return createElementToShorten(element, classId.asSingleFqName(), true)
+                }
+                // Allow using explicit import to overwrite members imported on-demand or in package
+                availableClassifier.importKind > ImportKind.DEFAULT_EXPLICIT && option == ShortenOption.SHORTEN_AND_IMPORT -> {
+                    return createElementToShorten(element, classId.asSingleFqName(), false)
+                }
             }
         }
-
-        val (mostTopLevelClassId, mostTopLevelQualifier) = allClassIds.zip(allQualifiers).last()
-        val availableClassifier = shorteningContext.findFirstClassifierInScopesByName(positionScopes, mostTopLevelClassId.shortClassName)
-
-        check(availableClassifier?.classId != mostTopLevelClassId) {
-            "If this condition were true, we would have exited from the loop on the last iteration. ClassId = $mostTopLevelClassId"
-        }
-
-        return if (availableClassifier == null || availableClassifier.isFromStarOrPackageImport) {
-            ShortenQualifier(mostTopLevelQualifier, mostTopLevelClassId.asSingleFqName())
-        } else {
-            findFakePackageToShorten(mostTopLevelQualifier)
-        }
+        return findFakePackageToShortenFn(allElements.last())
     }
 
+    private fun ClassId.toRegularClass() =
```

Comment:
It seems like you purposefully ignore typealiases here (by casting to FirRegularClassSymbol) - is this neccessary?

Reply:
Ah! Initially I thought type alias can't be imported. I was wrong and there is no need to do this cast. Thanks!